### ISSUE TITLE
Large merge transactions hang

### ISSUE THREAD

**Issue by technige** [neo4j/neo4j-python-driver#47]
For details see https://github.com/nigelsmall/py2neo/issues/484

**Comment by saoirse-ronan** [comment]
With py2neo V3 B2 and neo4J 3.0.0 RC1, we are getting less hangs with Bolt but it still hangs on occasion.  If we disable Bolt transactions, the script runs to completion without complaint. When our script hangs, the stack trace shows its hanging during the commit:

```
python3 neo-ingestion-rate.py
Number of batches: 8
Entity transactions per batch: 6144
Merge entities: 2016-04-22 13:24:18.449533
^CTraceback (most recent call last):
  File "neo-ingestion-rate.py", line 143, in <module>
    transaction.commit()
  File "/usr/lib/python3.4/site-packages/py2neo/database/__init__.py", line 999, in commit
    self._post(commit=True)
  File "/usr/lib/python3.4/site-packages/py2neo/database/__init__.py", line 1228, in _post
    self.finish()
  File "/usr/lib/python3.4/site-packages/py2neo/database/__init__.py", line 1233, in finish
    self._sync()
  File "/usr/lib/python3.4/site-packages/py2neo/database/__init__.py", line 1223, in _sync
    connection.fetch()
  File "/usr/lib/python3.4/site-packages/py2neo/packages/neo4j/v1/connection.py", line 292, in fetch
    raw.writelines(self.channel.chunk_reader())
  File "/usr/lib/python3.4/site-packages/py2neo/packages/neo4j/v1/connection.py", line 167, in chunk_reader
    chunk_header = self._recv(2)
  File "/usr/lib/python3.4/site-packages/py2neo/packages/neo4j/v1/connection.py", line 145, in _recv
    b = self.socket.recv(8192)
  File "/usr/lib64/python3.4/ssl.py", line 736, in recv
    return self.read(buflen)
  File "/usr/lib64/python3.4/ssl.py", line 625, in read
    v = self._sslobj.read(len or 1024)
KeyboardInterrupt
```

**Comment by saoirse-ronan** [comment]
Bolt transactions work properly now with the released version of Neo4J 3.0.0.   No more hangs.

**Comment by technige** [comment]
Closing for now as problem has not been seen in final release. Can be reopened if this changes.

**Comment by mikayla-grace** [comment]
We're seeing hangs with the Bolt protocol.  It consistently happens with Neo4J 3.0.3 and py2neo 3 but only after loading several million nodes and relationships.  When it does hang, other Neo4J activities can continue such as backups or browser queries.  There is nothing referencing the hang in the Neo4J logs.  If we start another script, it hangs when we call Graph().  It does not timeout, it just sits there for hours until we kill the script or restart Neo4J.    In the mean-time we switched to a HTTP connection and so far we have not seen any hangs.   Perhaps the problem is in Neo4J instead of py2neo.

**Comment by jexp** [comment]
@mikayla-grace can you provide a thread-dump with `kill -3 <pid> or`jstack <pid>`

**Comment by mikayla-grace** [comment]
Unfortunately it takes up to 2 days before the hang occurs and we can't afford any more downtime populating our graph database.  However, we do have a smaller host with less memory.  We'll see if we can reproduce the problem there without running out of memory and get a stack dump.  We'll post here if we can reproduce the Bolt protocol hang.

In the mean-time, since we switched to HTTP protocol, most of our ingestion scripts appear to be working but one is failing with a HTTP exception:

```
2016-06-30 22:31:09.676+0000 WARN  badMessage: java.lang.IllegalStateException: too much data after closed for HttpChannelOverHttp@185b8e50{r=1,c=false,a=IDLE,uri=-}
```

We also received yet another exception just now since we switched to the HTTP protocol:

```
File "/usr/lib/python3.5/site-packages/py2neo/database/__init__.py", line 1152, in _post
raise GraphError.hydrate(raw_error)
py2neo.database.status.DatabaseError: File has been unmapped: /<path>/data/databases/graph.db/neostore.counts.db.a
```

Environment, Fedora Linux, 12 cores, 32GB of memory.  Java 1.8.0_92-b14.

**Comment by jakewins** [comment]
For that last error, could you check if there is a stack trace in `logs/debug.log` or `logs/neo4j.log` with more details on that error?

Also, as an alternative to providing thread dumps, could you give a bit more detail on what you are doing, and we can try and replicate it on our end? For example:
- What do the queries look like
- Maybe a small python code sample of what you do in the client
- How many times do you run the queries before hitting issues
- What does `uname -r` say
- What python version, and what specific py2neo version? 

And so on; anything you think could be helpful to try and replicate it.

**Comment by mikayla-grace** [comment]
We're ingesting data from a variety of public sources using a Python class we wrote.  We're using Python 3.5 and Neo4J 3.0.3 Enterprise and py2neo 3.0.0, we just upgraded to py2neo 3.1.0.  We need to ingest around 25 million records and the scripts have failed repeatedly with either a hang with Bolt, or we get a 404 for HTTP, or we get a graph.hydrate exception with HTTP.

The failures are during the ingestion process.  We have not queried much because we have been unable to create a stable graph database.  We use py2neo merge() with transactions.  At first we thought the transactions might be too big so we cut back from 1000 transactions before a commit to 64 now.

We typically ingest nearly 2.5 million records from 4 different scripts ingesting from 4 different public data sources.  With Bolt, the scripts typically hang in the same or nearly the same place usually around 20 hours or so after the ingestion starts.    With HTTP, one or more scripts will typically fail with a graph.hydrate or HTTP 404 exception usually within a few hours.

The code is pretty straightforward:

```
    node=Node(label)
    <add a few properties to the Node>
    <check the transaction limit, if more than 64, commit>
    self.transaction.merge(node,label,{ key: value})
```

Here is our kernel: 4.5.7-200.fc23.x86_64+debug.

We have a 1.4MB log file.  One exception is:

```
2016-06-29 18:24:30.755+0000 ERROR [o.n.b.v.r.i.ErrorReporter] Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference f9f73c55-1502-4ed1-b644-9185eff49375.
2016-06-29 18:24:30.756+0000 ERROR [o.n.b.v.r.i.ErrorReporter] Client triggered an unexpected error [UnknownError]: null, reference f9f73c55-1502-4ed1-b644-9185eff49375.
java.lang.NullPointerException
        at org.neo4j.bolt.v1.transport.ChunkedOutput.ensure(ChunkedOutput.java:156)
        at org.neo4j.bolt.v1.transport.ChunkedOutput.writeShort(ChunkedOutput.java:90)
        at org.neo4j.bolt.v1.packstream.PackStream$Packer.packStructHeader(PackStream.java:304)
        at org.neo4j.bolt.v1.messaging.PackStreamMessageFormatV1$Writer.handleRecordMessage(PackStreamMessageFormatV1.java:132)
        at org.neo4j.bolt.v1.messaging.msgprocess.RecordStreamCallback$1.visit(RecordStreamCallback.java:47)
        at org.neo4j.bolt.v1.runtime.internal.CypherAdapterStream.lambda$accept$9(CypherAdapterStream.java:68)
        at org.neo4j.cypher.internal.compatibility.ExecutionResultWrapperFor3_0$$anon$5.visit(CompatibilityFor3_0.scala:410)
        at org.neo4j.cypher.internal.compiler.v3_0.helpers.RuntimeJavaValueConverter$feedIteratorToVisitable.accept(RuntimeJavaValueConverter.scala:54)
        at org.neo4j.cypher.internal.compiler.v3_0.PipeExecutionResult.accept(PipeExecutionResult.scala:109)
```

We also got a ' Discarded stale query from the query cache' perhaps expected after a log rotation.

We could make the scripts and the debug logs available privately to you if you prefer.  If so, send a private email to our temporary e-mail address at neophite at virginimage dot org.

**Comment by jexp** [comment]
Can you share the concrete code?

I'm not sure what py2neo's merge does under the hood.

I'd go with a plain unwind + merge statement (I presume you have the necessary constraint(s) in place).

Something like: (pass a list of size 10k to 100k  of dictionaries as "rows" parameter))

```
UNWIND {rows} as row
    MERGE (n:Label {id: row.id}) ON CREATE SET n += row
```

you can also spell out the individual properties in the SET clause

**Comment by mikayla-grace** [comment]
A subset of the concrete code is:

```
    self.graph=Graph(graphURL,user=graphUser,password=graphPassphrase,bolt=False)
    self.transaction=self.graph.begin()
    ...
  def mergeEntity(self,label,key,value,labels,**properties):
    '''
      Merge an entity graph node;  label is required, labels are optional.
    '''
    if value == None or value == '': return
    node=Node(label)
    if labels != None:
      if not isinstance(labels,list):
        node.add_label(labels)
      else:
        node.update_labels(labels)
    if "name" not in properties:
      node["name"]=value
    for _property in properties:
      node[_property]=properties[_property]
    node[key]=value
    name=label+'-'+key+'-'+value
    node['type']=self.slugify(label)
    node['id']=self.slugify(value)
    node['nameHash']=self.slugify(node["name"])
    if node["uuid"]!= False:
      node['uuid']=str(uuid.uuid5(uuid.NAMESPACE_DNS,name))
    node["source"]=self.source
    node['importDate']=datetime.datetime.utcnow().isoformat()+'Z'
    self.checkTransactionLimits('update','entity',name,node)
    self.transaction.merge(node,label,{ key: value})
    return(node)
...      
  self.transaction.commit()
```

A client looks like this:

```
  neoPhite.mergeEntity('Article',"id",_id,subType,
    url='http://mysite.com/'+entity['id'],
    name=neoPhite.canonicalProperty(entity,'title'),
    description=neoPhite.canonicalProperty(entity,'text'),
    title=neoPhite.canonicalProperty(entity,'title''),
   ...
```

Each script might have 10 to 40 different properties.

We're fairly invested in py2neo.  Our scripts were developed on a smaller set of data and it always seemed to work.  It only started to fail when we scaled.   Rather than completely rewriting our 20 plus scripts, maybe it would work if we reduced our dataset to batches of 100 records or so and run them as separate script instances.  If all else fails, we might revert to Neo4J version 2 series with fingers crossed that that will work.

**Comment by mikayla-grace** [comment]
We're now having some success with the py2neo 3.1 HTTP protocol ingesting large amounts of data into Neo4J 3.0.3.  Graph() does not accept a timeout parameter, it probably should.  Instead we set `http.socket_timeout` to a large value and our ingestion scripts have not thrown any exceptions.  If they do, we'll let you know.  Some of our scripts run for several days and it seems likely a socket timeout may throw spurious exceptions making it difficult to track.  Perhaps the Bolt transport is timing out as well on long runs.  However, we cannot find a way to set a timeout for Bolt.  Bolt works for us for any script that runs for less than 20 hours or so.  It hangs when a script runs for several days.

**Comment by technige** [comment]
Thanks for the detailed report. I'm going to be looking at timeout configuration across Bolt and all the drivers over the next few weeks since this issue has surfaced a few times in different places. Stay tuned for some improvements here.

**Comment by mikayla-grace** [comment]
Nigel, looking forward toward a patch to this problem.  In the mean-time, my scripts returned another hydrate exception.  Not sure its related:

```
Traceback (most recent call last):
  File "/knovo/knovo/Fusion/KnovoGraph.py", line 240, in commitTransaction
    self.transaction.commit()
  File "/usr/lib/python3.4/site-packages/py2neo/database/__init__.py", line 1022, in commit
    self._post(commit=True)
  File "/usr/lib/python3.4/site-packages/py2neo/database/__init__.py", line 1193, in _post
    raise GraphError.hydrate(raw_error)
py2neo.database.status.TransientError: ForsetiClient[5] can't acquire ExclusiveLock{owner=ForsetiClient[3]} on NODE(1552), because holders of that lock are waiting for ForsetiClient[5].
 Wait list:ExclusiveLock[
Client[3] waits for [5]]
```

Perhaps we just need to retry when this exception is thrown.

**Comment by mikayla-grace** [comment]
We thought our HTTP troubles were behind us when we set http.socket_timeout to 2592000 but our script just failed with a 404:

```
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/py2neo/database/http.py", line 203, in post
    response = self.__base.post(body, headers, **kwargs)
  File "/usr/lib/python3.4/site-packages/py2neo/packages/httpstream/http.py", line 984, in post
    return rq.submit(**kwargs)
  File "/usr/lib/python3.4/site-packages/py2neo/packages/httpstream/http.py", line 452, in submit
    return Response.wrap(http, uri, self, rs, **response_kwargs)
  File "/usr/lib/python3.4/site-packages/py2neo/packages/httpstream/http.py", line 489, in wrap
    raise inst
py2neo.packages.httpstream.http.ClientError: 404 Not Found
```

**Comment by jexp** [comment]
@nigelsmall does this funcitionality use cypher under the hood? I presume it must when it's usable via bolt?

**Comment by jexp** [comment]
@mikayla-grace it's not so much the number of transactions but the number of entities per transaction. How much memory (heap) is your server configured with (neo4j-wrapper.conf)?

**Comment by mikayla-grace** [comment]
We use the default configuration which as you know is dynamically calculated based on available system resources.  We have a dedicated 32GB 12 core Linux host (only the OS and neo4J runs on the system) so one would think that 64 transactions before a commit is sufficiently small enough to not cause heap problems.  We also track the string length of each transaction, if the total length exceeds 1MB, the transaction is committed.  However, if you have a recommendation on heap size or transaction size, let us know and we'll try our scripts with your recommendations.

**Comment by jexp** [comment]
There is only automatic computation for the page-cache (i.e. 50% of total RAM).

JVM heap is allocated by java if not configured, but  it usually only takes a small amount of the RAM as heap. So I presume the heap is just too small.

You can check the actual size of the heap and page cache in debug.log

**Comment by mikayla-grace** [comment]
Here is what we found in debug.log:

```
 System memory information:
 Total Physical memory: 31.27 GB
 Free Physical memory: 188.66 MB
 Committed virtual memory: 11.99 GB
 Total swap space: 15.69 GB
 Free swap space: 15.16 GB
 JVM memory information:
 Free  memory: 390.97 MB
 Total memory: 502.00 MB
 Max   memory: 7.82 GB
 Garbage Collector: G1 Young Generation: [G1 Eden Space, G1 Survivor Space]
 Garbage Collector: G1 Old Generation: [G1 Eden Space, G1 Survivor Space, G1 Old Gen]
 Memory Pool: Code Cache (Non-heap memory): committed=4.44 MB, used=4.41 MB, max=240.00 MB, threshold=0.00 B
 Memory Pool: Metaspace (Non-heap memory): committed=13.25 MB, used=12.55 MB, max=-1.00 B, threshold=0.00 B
 Memory Pool: Compressed Class Space (Non-heap memory): committed=1.75 MB, used=1.55 MB, max=1.00 GB, threshold=0.00 B
 Memory Pool: G1 Eden Space (Heap memory): committed=34.00 MB, used=14.00 MB, max=-1.00 B, threshold=?
 Memory Pool: G1 Survivor Space (Heap memory): committed=4.00 MB, used=4.00 MB, max=-1.00 B, threshold=?
 Memory Pool: G1 Old Gen (Heap memory): committed=464.00 MB, used=92.40 MB, max=7.82 GB, threshold=0.00 B
 Operating system information:
 Operating System: Linux; version: 4.5.6-200.fc23.x86_64+debug; arch: amd64; cpus: 12
 Max number of file descriptors: 327680
 Number of open file descriptors: 101
 Byte order: LITTLE_ENDIAN
 Local timezone: America/New_York
 JVM information:
 VM Name: OpenJDK 64-Bit Server VM
 VM Vendor: Oracle Corporation
 VM Version: 25.92-b14
 JIT compiler: HotSpot 64-Bit Tiered Compilers
 VM Arguments: [-XX:+UseG1GC, -XX:-OmitStackTraceInFastThrow, -XX:hashCode=5, -XX:+AlwaysPreTouch, -XX:+UnlockExperimentalVMOptions, -XX:+TrustFinalNonStaticFields, -XX:+DisableExplicitGC, -Dunsupported.dbms.udc.source=tarball, -Dfile.encoding=UTF-8]
 Java classpath:
  [bootstrap] /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.92-1.b14.fc23.x86_64/jre/lib/jsse.jar
  [bootstrap] /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.92-1.b14.fc23.x86_64/jre/lib/charsets.jar
```

Recommendations for configuring memory is welcome.  However, would memory problems cause the Bolt protocol to hang, exclusive lock exceptions, or HTTP post status of 404?

**Comment by dfeinzeig** [comment]
@mikayla-grace , did you figure this out?

I also have a script which is failing with

```
java.lang.IllegalStateException: too much data after closed for HttpChannelOverHttp
```

**Comment by mikayla-grace** [comment]
> did you figure this out?

We continue to have numerous problems with Neo4J enterprise 3.0.4 and py2neo 3.1.  Just today we got another deadlock during the ingestion.  We had to restart Neo4J and restart our ingestion scripts to recover.   We also have slow query responses.  As previously mentioned, we can ingest 20 million articles into Solr within 2 days.  With Neo4J and py2neo, these same articles will take close to a year.  We've been ingesting for months now and we're up to 4 million articles.  It seems that ingestion is slowing down as the number of articles increases.  We keep hoping that as each new release of Neo4J and py2neo are posted, that these problems will disappear.  If that does not happen soon, we will likely need to change to a different graph database.

**Comment by jexp** [comment]
Could you report in detail what you're doing, share you code, config an setup. 

Also raise a support ticket for your enterprise license. Neo4j can easily ingest 20m nodes within a few minutes. 

Von meinem iPhone gesendet

**Comment by mikayla-grace** [comment]
> Also raise a support ticket for your enterprise license. 

We have a start-up enterprise license.  Not sure support is covered under that license since it was free.

> Could you report in detail what you're doing, share your code, config an setup

We're using a Python class and an ingestion script to load PubMed citations into Neo4j with py2neo on a single 8 core Linux host with 32GB of memory and 6GB Sata.  We're ingesting about 6 thousand citations an hour.  Given we have 25 million citations to load,  it will take many months of continuous processing to finish.

In the past, while ingesting the citations, we've had Neo4J hang on multiple occasions.  The only way to get it working again was to restart Neo4J and then restart the ingestion scripts.  We switched from Bolt to HTTP and the hang mostly went away, although we still get them on occasion.  With HTTP we get hydrate errors due to deadlocks.  We wait 30 seconds and try the transaction commit again but it never works-- another exception is thrown.

You can review our Python scripts and reproduce the slow ingestion rate and maybe the deadlocks we're been getting:  http://urban-warrior.org/citations.tar.gz.

**Comment by ikwattro** [comment]
```
2016-06-30 22:31:09.676+0000 WARN  badMessage: java.lang.IllegalStateException: too much data after closed for HttpChannelOverHttp@185b8e50{r=1,c=false,a=IDLE,uri=-}
```

I have the same issue in two different setups.

First setup : 

Neo4j php client ingesting data under heavy load, sometimes 300-400 transactions per second.

It happens only when two requests in a row returns a ConstraintViolation exception.

Second setup, java process running http requests to an unmanaged extension

Note here that it never hangs, it is just logged in the server logs. I had one time a hang in 2.3 version

**Comment by jexp** [comment]
can you share more information about your data model
and the schema (indexes and constraints you have in place)

Thanks

**Comment by mikayla-grace** [comment]
While ingesting, we just got another exception: py2neo.database.status.TransientError: ForsetiClient[1] can't acquire ExclusiveLock{owner=ForsetiClient[2]} on NODE(2335), because holders of that lock are waiting for ForsetiClient[1].  We tried to commit again after 30 seconds and the second commit failed as well.

The data model revolves around citations which we call articles.  The article ID is our only unique constraint.  Each article has an ID, name, description, source, publication date, etc.  In addition we have other nodes that each article might be associated with.  They include author, grant, chemicals, and journals.  We first create the articles, authors, grants, chemicals and journal nodes.  Next we re-read the citations and create relationships from the articles to the associated authors, grants, chemicals, and journals.  We use merge() instead of create() because we so often need to restart the ingestion and this permits us to initially ingest the nodes as well as update them later using the same class.  We did try create() at some point and did not get the performance increase we were expecting.

**Comment by jexp** [comment]
But if you don't have any other constraints / indexes, it has to do a full
graph/label scan e.g. to find a single author, grant, chemical or journal
node, and that for each relationship that you're trying to create?

Please make sure that your query is using indexes for all node lookups with
EXPLAIN.

**Comment by mikayla-grace** [comment]
> it has to do a full graph/label scan

Kudos.  That made a significant difference.  That increased performance from 6,000 articles an hour to 75,000  (projected).  Let's see if that gets rid of the periodic Neo4J hangs.  Apologies for not spotting that problem ourselves.    We're benchmarking 300,000 articles now.  As soon as it completes, we'll  post the ingestion rate here.

**Comment by jexp** [comment]
Good luck.

How many relationships are created per article?

With a heap size of 6-8G you can create 1M relationships per transaction.

**Comment by mikayla-grace** [comment]
> How many relationships are created per article?

It varies of course but its likely no one article would have more than one hundred relationships.  Will have better metrics as soon as the benchmark completes.

**Comment by ikwattro** [comment]
@mikayla-grace maybe you should consider to make custom cypher queries for importing your data. With Cypher I have a write rate of approx 6k nodes, 27k rels per second (standard benchmark script I use)

**Comment by mikayla-grace** [comment]
We thought of that-- custom cypher queries, however, we use the same class for about 20 different public datasets we ingest with additional disparate datasets in the works.    We can also use this single elegant object-oriented Python class whenever we need to update already ingested nodes.   We have custom cypher queries on the back burner if all else fails.  Thanks.

**Comment by mikayla-grace** [comment]
@jexp.  That did it.  We ingested 750,000 nodes and relationships in 1 hour without any exceptions or hangs.  Apologies for not detecting this flaw earlier.  It is likely that all the exceptions and hangs we encountered in the past were related to searching un-indexed fields.  It also explains why our ingestion process got slower as we ingested more and more data.   If I may, we'll use the start-up card.  Always too much to do-- always too little time to do things right.
